src/TheFox/Smtp/Server.php

```diff
@@ -252,4 +262,28 @@ public function mailNew($from, $rcpt, $mail){
 		$this->eventExecute(Event::TRIGGER_MAIL_NEW, array($from, $rcpt, $mail));
 	}
 	
+	/**
+	 * Execute authentication events
+	 * 
+	 * All authentication events must return true for authentication to be successful
+	 * 
+	 */
+	public function authenticateUser($method, $credentials = array()){
+		$authenticated = false;
+		$args = array($method, $credentials);
+		
+		foreach($this->events as $eventId => $event){
+			if($event->getTrigger() == Event::TRIGGER_AUTH_ATTEMPT){
+				if($event->execute($args)){
+					$authenticated = true;
+				}
+				else{
+					$authenticated = false;
```

Comment:
My thinking was that if you set more then one `Event::TRIGGER_AUTH_ATTEMPT` event,  then all should return true. Any failed authentication attempts would be handled inside the callback. 

This would allow you to attempt to authenticate against two different systems and if one of them had blocked the user then the authentication process would fail.